Fix data loss issue in `combine_echodata`

b-reyes opened the pull request:
This PR addresses #822. This is done by first creating a mapping between uniform chunks and the initial starting chunks. Then a Dask Lock is assigned to each of the initial starting chunks so that no starting chunks are written to the uniform chunk at the same time. To illustrate the approach, consider the following simplified example:

Say we have three files each with the variable 'back_r' that contain the following values (these would be the starting chunks):
```
file 1: back_r = [0,1,2]

file 2: back_r = [3,4,5]

file 3: back_r = [6,7]
```

We then want to combine all of these `back_r` variables into `back_r_combined = [0,1,2,3,4,5,6,7]` with uniform chunk size 2.

Thus, the chunks would be as follows: 
```
chunk 1: [0,1]

chunk 2: [2,3]

chunk 3: [4,5]

chunk 4: [6,7]
```

For all chunks besides `chunk 2`, we can safely write the processes in parallel. However, we see that `chunk 2` contains data from file 1 and file 2. Thus, two different processes will attempt to write to `chunk 2` at the same time and data corruption will likely occur. To remedy this, we can assign a lock name to each write to a uniform chunk:
```
data write [0,1]  will be given lock name = "lock1"

data write [2] and [3]  will be given lock name = "lock2"

data write [4,5]  will be given lock name = "lock3"

data write [6,7]  will be given lock name = "lock4"
```

We then use Dask [Lock](https://distributed.dask.org/en/stable/api.html?highlight=lock#distributed.Lock) with the establish lock names and we can prevent chunks being written to at the same time.

b-reyes commented:
@lsetiawan I was unable to find a sufficient way to check if a chunk was correctly written using zarr (the sanity check we discussed). The reason for this is that zarr only allows you to view if a chunk has been initialized. This basically amounts to checking the directory and seeing if a chunk file exists. Thus, if one process partially fills a chunk, in zarr's perspective that chunk has been initialized. I could not find a sufficient way to check for partially filled chunks without loading the chunk into memory. Perhaps we can look into this more in the future.

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -387,6 +393,17 @@ def combine_echodata(
+    if client is None:
+        client = Client()  # create client with local scheduler
+        print(f"Client dashboard link: {client.dashboard_link}")
+    else:
+
+        if isinstance(client, Client):
+            print(f"Client dashboard link: {client.dashboard_link}")
+        else:
+            raise TypeError("The input client is not of type dask.distributed.Client!")
```
```suggestion
    elif isinstance(client, Client):
        print(f"Client dashboard link: {client.dashboard_link}")
    else:
        raise TypeError(f"The input client is not of type {type(Client)}!")
```

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -311,6 +313,8 @@ def combine_echodata(
     storage_options: Optional[dict]
         Any additional parameters for the storage
         backend (ignored for local paths)
+    client: Optional[dask.distributed.Client]
```
`dask` should be imported above.

lsetiawan commented:
Please add `distributed` to `requirements.txt`

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -387,6 +394,24 @@ def combine_echodata(
     """
     # TODO: change PR #297 reference to a link in our documentation
 
+    # check the client input and print dashboard link
+    if client is None:
+
+        # set flag specifying that a client was created
+        client_created = True
+
+        client = Client()  # create client with local scheduler
+        print(f"Client dashboard link: {client.dashboard_link}")
+    else:
+
+        # set flag specifying that a client was not created
+        client_created = False
+
+        if isinstance(client, Client):
+            print(f"Client dashboard link: {client.dashboard_link}")
+        else:
+            raise TypeError("The input client is not of type dask.distributed.Client!")
```
```suggestion
    # set flag specifying that a client was not created
    client_created = False

    # check the client input and print dashboard link
    if client is None:
        # set flag specifying that a client was created
        client_created = True

        client = Client()  # create client with local scheduler
        print(f"Client dashboard link: {client.dashboard_link}")
    elif isinstance(client, Client):
        print(f"Client dashboard link: {client.dashboard_link}")
    else:
        raise TypeError("The input client is not of type dask.distributed.Client!")
```

b-reyes commented:
@lsetiawan In addition to the above fixes, I also modified `test_combine_echodata` so that it constructs the expected combined Dataset  that `combine_echodata` should produce. A few small remarks 
- I removed the `"ek60_reversed_ping_time_test_data"` because this data is already being tested. More importantly, the comparison method in `test_combine_echodata` does not correct the reversed ping times in the test dataset, so all values are incorrectly compared. 
- I am currently not using the `EK80` test files for `test_combine_echodata` because they have attributes that are different (which is behavior that is not permitted in the current version of `combine_echodata`)

lsetiawan commented:
@b-reyes I tried using the echopype example of the solar eclipse as my test dataset for this, and here's what I found. Everything seem to convert and combined as expected, it seems like memory usage did not spike very much during combine. It's really cool to see the dask cranking away during combine! 😄 I did run to a TypeError however when I'm trying to act on the combined echodata object:

```python
ds_Sv = ep.calibrate.compute_Sv(combined)
```
results in 

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [21], in <cell line: 1>()
----> 1 ds_Sv = ep.calibrate.compute_Sv(combined)

File ~/GitRepos/GitHub/echopype/echopype/calibrate/api.py:202, in compute_Sv(echodata, **kwargs)
    116 def compute_Sv(echodata: EchoData, **kwargs) -> xr.Dataset:
    117     """
    118     Compute volume backscattering strength (Sv) from raw data.
    119 
   (...)
    200     it must be set using `EchoData.update_platform()`.
    201     """
--> 202     return _compute_cal(cal_type="Sv", echodata=echodata, **kwargs)

File ~/GitRepos/GitHub/echopype/echopype/calibrate/api.py:54, in _compute_cal(cal_type, echodata, env_params, cal_params, waveform_mode, encode_mode)
     48         logger.warning(
     49             "This sonar model only record data as power or power/angle samples "
     50             "(encode_mode='power'). Calibration will be done on the power samples.",
     51         )
     53 # Set up calibration object
---> 54 cal_obj = CALIBRATOR[echodata.sonar_model](
     55     echodata,
     56     env_params=env_params,
     57     cal_params=cal_params,
     58     waveform_mode=waveform_mode,
     59     encode_mode=encode_mode,
     60 )
     62 def add_attrs(cal_type, ds):
     63     """Add attributes to backscattering strength dataset.
     64     cal_type: Sv or TS
     65     """

File ~/GitRepos/GitHub/echopype/echopype/calibrate/calibrate_ek.py:240, in CalibrateEK60.__init__(self, echodata, env_params, cal_params, **kwargs)
    238 if cal_params is None:
    239     cal_params = {}
--> 240 self.get_cal_params(cal_params, waveform_mode="CW", encode_mode="power")
    242 # default to CW mode recorded as power samples
    243 self.compute_range_meter(waveform_mode="CW", encode_mode="power")

File ~/GitRepos/GitHub/echopype/echopype/calibrate/calibrate_ek.py:127, in CalibrateEK.get_cal_params(self, cal_params, waveform_mode, encode_mode)
    121     params_from_vend = ["sa_correction", "gain_correction"]
    122     for p in params_from_vend:
    123         # substitute if None in user input
    124         self.cal_params[p] = (
    125             cal_params[p]
    126             if p in cal_params
--> 127             else self._get_vend_cal_params_power(p, waveform_mode=waveform_mode)
    128         )
    130 # Other params
    131 self.cal_params["equivalent_beam_angle"] = (
    132     cal_params["equivalent_beam_angle"]
    133     if "equivalent_beam_angle" in cal_params
    134     else beam["equivalent_beam_angle"]
    135 )

File ~/GitRepos/GitHub/echopype/echopype/calibrate/calibrate_ek.py:97, in CalibrateEK._get_vend_cal_params_power(self, param, waveform_mode)
     90 da_param = (
     91     ds_vend[param][relevant_indexes]
     92     .expand_dims(dim={"ping_time": idxmin["ping_time"]})  # expand dims for direct indexing
     93     .sortby(idxmin.channel)  # sortby in case channel sequence different in vendor and beam
     94 )
     96 # Select corresponding index and clean up the original nan elements
---> 97 da_param = da_param.isel(pulse_length_bin=idxmin, drop=True)
     98 return da_param.where(~transmit_isnull, np.nan)

File /opt/conda/envs/ep39/lib/python3.9/site-packages/xarray/core/dataarray.py:1197, in DataArray.isel(self, indexers, drop, missing_dims, **indexers_kwargs)
   1194 indexers = either_dict_or_kwargs(indexers, indexers_kwargs, "isel")
   1196 if any(is_fancy_indexer(idx) for idx in indexers.values()):
-> 1197     ds = self._to_temp_dataset()._isel_fancy(
   1198         indexers, drop=drop, missing_dims=missing_dims
   1199     )
   1200     return self._from_temp_dataset(ds)
   1202 # Much faster algorithm for when all indexers are ints, slices, one-dimensional
   1203 # lists, or zero or one-dimensional np.ndarray's

File /opt/conda/envs/ep39/lib/python3.9/site-packages/xarray/core/dataset.py:2416, in Dataset._isel_fancy(self, indexers, drop, missing_dims)
   2414         indexes[name] = new_index
   2415 elif var_indexers:
-> 2416     new_var = var.isel(indexers=var_indexers)
   2417 else:
   2418     new_var = var.copy(deep=False)

File /opt/conda/envs/ep39/lib/python3.9/site-packages/xarray/core/variable.py:1186, in Variable.isel(self, indexers, missing_dims, **indexers_kwargs)
   1183 indexers = drop_dims_from_indexers(indexers, self.dims, missing_dims)
   1185 key = tuple(indexers.get(dim, slice(None)) for dim in self.dims)
-> 1186 return self[key]

File /opt/conda/envs/ep39/lib/python3.9/site-packages/xarray/core/variable.py:778, in Variable.__getitem__(self, key)
    765 def __getitem__(self: T_Variable, key) -> T_Variable:
    766     """Return a new Variable object whose contents are consistent with
    767     getting the provided key from the underlying data.
    768 
   (...)
    776     array `x.values` directly.
    777     """
--> 778     dims, indexer, new_order = self._broadcast_indexes(key)
    779     data = as_indexable(self._data)[indexer]
    780     if new_order:

File /opt/conda/envs/ep39/lib/python3.9/site-packages/xarray/core/variable.py:634, in Variable._broadcast_indexes(self, key)
    632 if isinstance(k, Variable):
    633     if len(k.dims) > 1:
--> 634         return self._broadcast_indexes_vectorized(key)
    635     dims.append(k.dims[0])
    636 elif not isinstance(k, integer_types):

File /opt/conda/envs/ep39/lib/python3.9/site-packages/xarray/core/variable.py:763, in Variable._broadcast_indexes_vectorized(self, key)
    760 else:
    761     new_order = None
--> 763 return out_dims, VectorizedIndexer(tuple(out_key)), new_order

File /opt/conda/envs/ep39/lib/python3.9/site-packages/xarray/core/indexing.py:326, in VectorizedIndexer.__init__(self, key)
    324         k = np.asarray(k, dtype=np.int64)
    325     else:
--> 326         raise TypeError(
    327             f"unexpected indexer type for {type(self).__name__}: {k!r}"
    328         )
    329     new_key.append(k)
    331 super().__init__(new_key)

TypeError: unexpected indexer type for VectorizedIndexer: dask.array<transpose, shape=(109482, 3), dtype=int64, chunksize=(1000, 1), chunktype=numpy.ndarray>
```
I'm not sure if this is a `compute_Sv` function issue or a data issue. If I remember correctly, `compute_Sv` is able to work with dask array in the past.. @leewujung correct me if I'm wrong. If it's a data issue, then something is going on during combine, if it's a `compute_Sv` issue then we need to investigate further in a different PR.

Either way, I was thinking that it would be great to have an `EchoData.persist()` and `Echodata.compute()` to load in the data into the local memory or workers memory.

This is my initial finds for this PR, I'll dig more into the code throughout the day... bare with me... might be slow 😛

lsetiawan commented:
Okay. After some investigating I think this issue lies within the `compute_Sv` function. I tried passing echodata from `open_converted(chunks='auto')` vs `open_converted()`... The latter works with no problem since it's using the xarray builtin lazy arrays rather than dask arrays. I thought things worked with dask arrays in `compute_Sv`, but I guess not.

b-reyes commented:
@lsetiawan I did some searching and it appears the issue is with the use of `isel`. See the xarray issue [2511](https://github.com/pydata/xarray/issues/2511) and the PR [5873](https://github.com/pydata/xarray/pull/5873).

leewujung commented:
> If I remember correctly, `compute_Sv` is able to work with dask array in the past.. @leewujung correct me if I'm wrong.

[This was the PR](https://github.com/OSOceanAcoustics/echopype/pull/697) I submitted to make `compute_Sv` work in a delayed array -- I thought we tried that? From the above it seems that dask is expecting indexer of a certain type?

lsetiawan commented:
> I did some searching and it appears the issue is with the use of isel. See the xarray issue https://github.com/pydata/xarray/issues/2511 and the PR https://github.com/pydata/xarray/pull/5873.

Your suggestion with `sel` instead of `isel` works! I'll just need to check that it works for both cases with dask arrays or normal arrays.

> https://github.com/OSOceanAcoustics/echopype/pull/697 I submitted to make compute_Sv work in a delayed array -- I thought we tried that? From the above it seems that dask is expecting indexer of a certain type?

I thought we did also, but seems like we didn't. Anyhow, I think the small fix mention above seems promising.

@b-reyes I also found a small bug where if the combined zarr already exists, I can't combine again. I think having an `overwrite` option would be great. If user doesn't specify the overwrite option, maybe give them a warning that the file exists already.

![Screenshot from 2022-10-04 13-28-11](https://user-images.githubusercontent.com/17802172/193922531-7bdf9a5f-56e1-4b32-be16-fb7567475e90.png)

b-reyes commented:
> @b-reyes I also found a small bug where if the combined zarr already exists, I can't combine again. I think having an overwrite option would be great. If user doesn't specify the overwrite option, maybe give them a warning that the file exists already.

@lsetiawan thank you for bringing this to my attention. I now include an overwrite option as you suggested. Instead of a warning, I think we should raise a `RuntimeError` if the file exists and `overwrite=False`. I think it is better to raise an error because we need a path to write to in order to do anything useful in `ZarrCombine`.

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -387,8 +426,23 @@ def combine_echodata(
+        client = Client()  # create client with local scheduler
+        print(f"Client dashboard link: {client.dashboard_link}")
+    elif isinstance(client, Client):
+        print(f"Client dashboard link: {client.dashboard_link}")
```
Instead of print statements, this should now change to logger info statements.

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -54,13 +63,30 @@ def check_zarr_path(zarr_path: str, storage_options: Optional[dict]) -> str:
             save_path = path_obj.parent
 
-    return validate_output_path(
+    validated_path = validate_output_path(
```
Currently this validation will fail if I pass in path like `s3://zarr-bucket-test/combined.zarr`, this is because by the time save_path is passed into `validate_output_path` it becomes `s3:/zarr-bucket-test/` since you turned `zarr_path` into a Path object. You should just pass on `zarr_path` and let the `validate_output_path` do the work like so:

```python
    source_file = "combined_echodatas.zarr"
    if not zarr_path.strip("/").endswith(".zarr"):
        raise ValueError("The provided zarr_path input must have '.zarr' suffix!")

    validated_path = validate_output_path(
        source_file=source_file,
        engine="zarr",
        output_storage_options=storage_options,
        save_path=zarr_path,
    )
```

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -54,13 +63,30 @@ def check_zarr_path(zarr_path: str, storage_options: Optional[dict]) -> str:
+
+    if exists and not overwrite:
+        raise RuntimeError(
+            f"{zarr_path} already exists, please provide a different path" " or set overwrite=True."
```
```suggestion
            f"{zarr_path} already exists, please provide a different path or set overwrite=True."
```

lsetiawan commented:
@b-reyes I found another huge bug that should be fixed. When I point `zarr_path` directly to an s3 bucket zarr, `combine_echodata` doesn't work. Here's the code snippet I used.

```python
combined = ep.combine_echodata(
    ed_list,
    zarr_path="s3://zarr-bucket-test/combined.zarr",
    storage_options=dict(
        client_kwargs=dict(endpoint_url="http://localhost:9000/"),
        key="minioadmin",
        secret="minioadmin",
    ),
    client=client)
```

The above will pass all the `check_zarr_path` with my changes from https://github.com/OSOceanAcoustics/echopype/pull/824#discussion_r987400986, however, when your functions are trying to write/read the arrays I get an `ArrayNotFoundError`:

```python
---------------------------------------------------------------------------
ArrayNotFoundError                        Traceback (most recent call last)
Input In [14], in <cell line: 1>()
----> 1 combined = ep.combine_echodata(ed_list, zarr_path="s3://zarr-bucket-test/combined.zarr", storage_options=dict(
      2         client_kwargs=dict(endpoint_url="http://localhost:9000/"),
      3         key="minioadmin",
      4         secret="minioadmin",
      5     ), client=client)

File ~/GitRepos/GitHub/echopype/echopype/echodata/combine.py:416, in combine_echodata(echodatas, zarr_path, storage_options, client)
    413 comb = ZarrCombine()
    415 # combine all elements in echodatas by writing to a zarr store
--> 416 ed_comb = comb.combine(
    417     zarr_path,
    418     echodatas,
    419     storage_options=storage_options,
    420     sonar_model=sonar_model,
    421     echodata_filenames=echodata_filenames,
    422 )
    424 orchestrate_reverse_time_check(ed_comb, zarr_path, comb.possible_time_dims, storage_options)
    426 if client_created:
    427     # close client

File ~/GitRepos/GitHub/echopype/echopype/echodata/zarr_combine.py:1007, in ZarrCombine.combine(self, zarr_path, eds, storage_options, sonar_model, echodata_filenames)
   1004 self._append_provenance_attr_vars(zarr_path, storage_options=storage_options)
   1006 # change filenames numbering to range(len(eds))
-> 1007 self._modify_prov_filenames(zarr_path, len_eds=len(eds))
   1009 # open lazy loaded combined EchoData object
   1010 ed_combined = open_converted(
   1011     zarr_path, chunks={}, synchronizer=zarr.ThreadSynchronizer()
   1012 )  # TODO: is this appropriate for chunks?

File ~/GitRepos/GitHub/echopype/echopype/echodata/zarr_combine.py:910, in ZarrCombine._modify_prov_filenames(zarr_path, len_eds)
    895 """
    896 After the ``Provenance`` group has been constructed, the
    897 coordinate ``filenames`` will be filled with zeros. This
   (...)
    906     The number of ``EchoData`` objects being combined
    907 """
    909 # obtain the filenames zarr array
--> 910 zarr_filenames = zarr.open_array(zarr_path + "/Provenance/filenames", mode="r+")
    912 zarr_filenames[:] = np.arange(len_eds)

File /opt/conda/envs/ep39/lib/python3.9/site-packages/zarr/creation.py:556, in open_array(store, mode, shape, chunks, dtype, compressor, fill_value, order, synchronizer, filters, cache_metadata, cache_attrs, path, object_codec, chunk_store, storage_options, partial_decompress, write_empty_chunks, zarr_version, dimension_separator, **kwargs)
    554         if contains_group(store, path=path):
    555             raise ContainsGroupError(path)
--> 556         raise ArrayNotFoundError(path)
    558 elif mode == 'w':
    559     init_array(store, shape=shape, chunks=chunks, dtype=dtype,
    560                compressor=compressor, fill_value=fill_value,
    561                order=order, filters=filters, overwrite=True, path=path,
    562                object_codec=object_codec, chunk_store=chunk_store,
    563                dimension_separator=dimension_separator)

ArrayNotFoundError: array not found at path %r' ''

```

This happens because you're not passing the `storage_options` to your functions as you can see in line 910
`910 zarr_filenames = zarr.open_array(zarr_path + "/Provenance/filenames", mode="r+")`

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -2,7 +2,10 @@
 from typing import List, Optional, Tuple
```
```suggestion
from typing import List, Optional, Tuple, Union
```

lsetiawan replied:
```suggestion
from typing import List, Optional, Tuple, Dict, Any
```

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -13,7 +16,7 @@
 
 
-def check_zarr_path(zarr_path: str, storage_options: Optional[dict]) -> str:
+def check_zarr_path(zarr_path: str, storage_options: Optional[dict], overwrite: bool) -> str:
```
```suggestion
def check_zarr_path(zarr_path: Union[Path, str], storage_options: Optional[dict], overwrite: bool) -> str:
```

lsetiawan replied:
```suggestion
def check_zarr_path(zarr_path: str, storage_options: Dict = {}, overwrite: bool = False) -> str:
```

Few fixes here, probably good to set a default for `overwrite`. Also, seem like you did spread `storage_options` in line 68, there should be a default empty dictionary. `validate_output_path` doesn't take optional `storage_options`, so this shouldn't be optional as in `None` is able to be passed either. Whenever you have `Optional`, the default value should be `None` otherwise, that argument is required.

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -733,15 +1002,14 @@ def combine(
-
         # open lazy loaded combined EchoData object
         ed_combined = open_converted(
             zarr_path, chunks={}, synchronizer=zarr.ThreadSynchronizer()
```
```suggestion
            zarr_path, chunks={}, synchronizer=zarr.ThreadSynchronizer(), storage_options=storage_options
```

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -895,7 +895,9 @@ def _append_provenance_attr_vars(
     @staticmethod
-    def _modify_prov_filenames(zarr_path: str, len_eds: int) -> None:
+    def _modify_prov_filenames(
+        zarr_path: str, len_eds: int, storage_options: Optional[dict]
```
If you have `Optional` then you should set `storage_options=None`

b-reyes commented:
> The above will pass all the check_zarr_path with my changes from https://github.com/OSOceanAcoustics/echopype/pull/824#discussion_r987400986, however, when your functions are trying to write/read the arrays I get an ArrayNotFoundError

@lsetiawan based on our interaction yesterday and my inclusion of `storage_options` in `zarr.open_array`, this error seems like it has now been addressed. Please let me know if you experience any more issues once you start testing things out in the cloud.

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -25,6 +28,10 @@ def check_zarr_path(zarr_path: str, storage_options: Optional[dict]) -> str:
     storage_options: Optional[dict]
```
```suggestion
    storage_options: dict
```

lsetiawan commented on echopype/echodata/combine.py:
```diff
@@ -295,22 +315,30 @@ def orchestrate_reverse_time_check(
+        If True, will overwrite the zarr store specified by
+        ``zarr_path`` if it already exists, otherwise an error
+        will be returned if the file already exists.
     storage_options: Optional[dict]
```
```suggestion
    storage_options: dict
```

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -438,22 +450,194 @@ def _get_region(self, ds_ind: int, ds_dims: Set[Hashable]) -> Dict[str, slice]:
+        zarr_group: str
+            The name of the group of the zarr store
+            corresponding to the Datasets in ``ds_list``
+        region: Dict[str, slice]
```
```suggestion
        region: dict
```

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -438,22 +450,194 @@ def _get_region(self, ds_ind: int, ds_dims: Set[Hashable]) -> Dict[str, slice]:
+        region: Dict[str, slice]
+            Keys set as the dimension name and values as
+            the slice of the zarr portion to write to
+        storage_options: Optional[dict]
```
```suggestion
        storage_options: dict | None
```

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -578,6 +784,55 @@ def _append_const_to_zarr(
+        ds_list: List[xr.Dataset],
+        zarr_path: str,
+        zarr_group: str,
+        storage_options: Optional[dict],
```
```suggestion
        storage_options: Optional[dict] = None,
```

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -623,7 +895,9 @@ def _append_provenance_attr_vars(
     @staticmethod
-    def _modify_prov_filenames(zarr_path: str, len_eds: int) -> None:
+    def _modify_prov_filenames(
+        zarr_path: str, len_eds: int, storage_options: Optional[dict] = {}
```
```suggestion
        zarr_path: str, len_eds: int, storage_options: Optional[dict] = None
```

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -578,6 +784,55 @@ def _append_const_to_zarr(
+        zarr_group: str
+            The name of the group of the zarr store
+            corresponding to the Datasets in ``ds_list``
+        storage_options: Optional[dict]
```
```suggestion
        storage_options: dict | None
```

lsetiawan commented on echopype/echodata/zarr_combine.py:
```diff
@@ -636,10 +910,15 @@ def _modify_prov_filenames(zarr_path: str, len_eds: int) -> None:
             The full path of the final combined zarr store
         len_eds: int
             The number of ``EchoData`` objects being combined
+        storage_options: Optional[dict]
```
```suggestion
        storage_options: dict | None
```

b-reyes commented:
@lsetiawan I agree with you that I was incorrectly using `Optional` for `storage_options`, this should always be a required variable and have an empty dict for its value. I have hopefully changed all typing to reflect this. 

I am slightly puzzled about your changes to the doc strings. Based on your comments, it appears that you do not want typing types in the doc strings i.e. things like `Dict, Optional, etc.`. If this is the case, can you explain to me why this is? I don't necessarily disagree with it, but I would like some guidelines for my future documentation of echopype.

lsetiawan commented:
> I am slightly puzzled about your changes to the doc strings. Based on your comments, it appears that you do not want typing types in the doc strings i.e. things like Dict, Optional, etc.. If this is the case, can you explain to me why this is? I don't necessarily disagree with it, but I would like some guidelines for my future documentation of echopype.

I do like the straight type hints in there however, for users that are not familiar with type hinting, they will be reading it and might be puzzled. Type hints are a new feature in python, but the primitive types are a lot more familiar with the majority of people and it's much more readable. Take for example a docstring with a parameter of `start_count` and it can take `int` or `float`. A type hint would be `Union[int, float]`... however that's not very intuitive, it's great for IDE, but not a new python user that just wants to use the function based on the docstring so the more appropriate thing would be `start_count: int or float`. Anyways, type hints can get very cryptic real quick, and I don't think docstring is the right place for them.

b-reyes commented:
> Anyways, type hints can get very cryptic real quick, and I don't think docstring is the right place for them.

That is a fair point. I will go ahead and fix all doc strings that have them in it. 

To your point of user readability, I think I like `start_count: int or float`, rather than `start_count: int | float`. Do you mind if I use the format `start_count: int or float`? For an optional, this would look like `start_count: int or None`.

lsetiawan commented:
> Do you mind if I use the format start_count: int or float?

Yea that's the best I think also.

> For an optional, this would look like start_count: int or None

For this, based on numpydoc convention, it should be `int, optional`.

See:

![Screenshot from 2022-10-05 16-27-22](https://user-images.githubusercontent.com/17802172/194181829-d1b72138-3f3c-4528-a9f8-25bb68bc3621.png)

b-reyes commented:
> For this, based on numpydoc convention, it should be int, optional.

Thank you for that reference! Great, I will go with `int, optional`

lsetiawan commented:
Thanks! Sorry for all of these stylistic changes! I guess it's better cleaning them up now than later 😛

b-reyes commented:
> Thanks! Sorry for all of these stylistic changes! I guess it's better cleaning them up now than later 😛

No worries. I agree, it is better to change them now.

lsetiawan reviewed (approved):
After detailed testing found here: https://nbviewer.org/gist/lsetiawan/ebb3faed65e53a3188518d62dbe0968a I conclude that this `combine_echodata` doesn't have any data loss issue and it's able to combine a large amount of data with minimal impact on memory consumption/spikes. 

The example notebook that I tested on converted `318` EK60 Raw Files from OOI in **16min** and then combined those files in about **15min**. These times may be due to limitation in my CPU clock speed. The memory consumption in my machine never blew up and at the end I was able to explore **133GB** of data with ease.

AWESOME WORK @b-reyes! This PR is ready for merging. Please put your testing results in this PR for the Hake data.

b-reyes commented:
@lsetiawan thank you very much for investigating this PR and testing out a large amount of data files! I am glad to hear that memory consumption stayed steady and the runtime for the combination of the files was small. 

As @lsetiawan mentioned, I have also tested out Hake data. Specifically, I tested out `107` files within `s3://ncei-wcsd-archive/data/raw/Bell_M._Shimada/SH1707/EK60`. It took roughly 7 minutes to convert the data and 5 minutes to combine them. The notebook to run this can be found here: https://nbviewer.org/gist/b-reyes/77bbede005f266509618d1430ae8f33e, please be sure to provide a `Client` to `combine_echodata` before you run this!
